### PR TITLE
[SPARK-50144][SS] Address the limitation of metrics calculation with DSv1 streaming sources

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to make the majority of DSv1 streaming sources to carry over stream information from logical plan to physical plan, so that Spark can use the same approach of collecting metrics for DSv2 streaming sources.

To achieve this, this PR introduces two new traits which are matching with logical node and physical node respectively:

* Logical plan: StreamSourceAwareLogicalPlan
* Physical plan: StreamSourceAwareSparkPlan

All streaming DSv1 sources are expected to produce the logical plan in getBatch(), which has leaf node(s) implementing StreamSourceAwareLogicalPlan. For built-in DSv1 streaming sources (and external sources using `SparkSession/SQLContext.internalCreateDataFrame`), they are mostly using one or multiple of nodes:

* LogicalRDD
* LocalRelation
* LogicalRelation

Physical planning with LogicalRelation will be covered via either 1) RowDataSourceScanExec or 2) FileSourceScanExec. It may not cover all possible types of relations, but it's uneasy for both data source developers and users to extend the Spark planner to handle the additional case in physical planning. Furthermore, DSv2 is the standard interface for streaming sources for external sources. For others they are covered with static nodes, LogicalRDD -> RDDScanExec, LocalRelation -> LocalTableScanExec.

This PR updates the progress reporter to collect the metrics from the nodes implementing StreamSourceAwareSparkPlan, which can be used across DSv1 and DSv2. To avoid regression, the progress reporter will check whether the executed plan has all streams we want to capture; if the executed plan misses some stream(s), the progress reporter will fall back to the old way.

### Why are the changes needed?

For DSv2 data sources, the source nodes in the executed plan are always MicroBatchScanExec, and these nodes contain the stream information.

But for DSv1 data sources, the logical node and the physical node representing the scan of the source are technically arbitrary (any logical node and any physical node), hence Spark makes an assumption that the leaf nodes for initial logical plan <=> logical plan for batch N <=> physical plan for batch N are the same so that we can associate these nodes. This is fragile and we have non-trivial number of reports of broken metric.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Modified tests - some tests were based on limitation and these tests are fixed with this PR.

e.g. org.apache.spark.sql.streaming.TriggerAvailableNowSuite

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala [hunk 1]
```diff
@@ -170,10 +173,16 @@ class TriggerAvailableNowSuite extends FileStreamSourceTest {
...
+        val expectedNumBatches = if (!useWrapper && !supportTriggerAvailableNow) {
+          // Spark will decide to fall back to SingleBatchExecutor.
+          1
+        } else {
+          3
+        }
+
         try {
           assert(q.awaitTermination(streamingTimeout.toMillis))
-          // only one batch has data in both sources, thus counted, see SPARK-24050
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:175]: This change of test represents that this PR fixes the limitation.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala [hunk 2]
```diff
@@ -193,8 +202,7 @@ class TriggerAvailableNowSuite extends FileStreamSourceTest {
         val q2 = startQuery()
         try {
           assert(q2.awaitTermination(streamingTimeout.toMillis))
-          // only one batch has data in both sources, thus counted, see SPARK-24050
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TriggerAvailableNowSuite.scala:196]: Here as well

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @brkyvz @viirya Appreciated for reviewing. Thanks in advance.
cc. @cloud-fan as well, since this PR changes the co-used logical/physical nodes between batch and streaming sources.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/StreamSourceAwareSparkPlan.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.connector.read.streaming.SparkDataStream
+
+/**
+ * This trait is a mixin for source physical nodes to represent the stream. This is required to the
+ * physical nodes which is transformed from source logical nodes inheriting
+ * [[org.apache.spark.sql.catalyst.plans.logical.StreamSourceAwareLogicalPlan]].
+ */
+trait StreamSourceAwareSparkPlan extends SparkPlan {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/StreamSourceAwareSparkPlan.scala:26]: shall we mention that it must report `numOutputRows` metric?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/StreamSourceAwareSparkPlan.scala:26]: Ah OK, good point.

**Review by cloud-fan — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @brkyvz @viirya Friendly reminder. I'll merge the change tomorrow if there is no outstanding comment.

**Comment by HeartSaVioR** [comment]
Haven't got feedback so far. Thanks! Merging to master.

**Comment by HeartSaVioR** [comment]
(ignore, check for Github service health)

**Pull request closed without merging**